packages/controllers/src/plugins/PluginController.test.ts

```diff
@@ -439,4 +490,83 @@ describe('PluginController Controller', () => {
       }),
     );
   });
+
+  it('can handle an error event on the controller messenger', async () => {
+    const controllerMessenger = new ControllerMessenger<
+      PluginControllerActions,
+      ErrorMessageEvent
+    >();
+    const messenger = controllerMessenger.getRestricted({
+      name: 'PluginController',
+      allowedEvents: ['ServiceMessenger:unhandledError'],
+    });
+
+    const workerExecutionEnvironment = new WebWorkerExecutionEnvironmentService(
+      {
+        messenger,
+        setupPluginProvider: jest.fn(),
+        workerUrl: new URL(URL.createObjectURL(new Blob([workerCode]))),
+      },
+    );
+    const pluginController = new PluginController({
+      terminateAllPlugins: workerExecutionEnvironment.terminateAllPlugins.bind(
+        workerExecutionEnvironment,
+      ),
+      terminatePlugin: workerExecutionEnvironment.terminatePlugin.bind(
+        workerExecutionEnvironment,
+      ),
+      executePlugin: workerExecutionEnvironment.executePlugin.bind(
+        workerExecutionEnvironment,
+      ),
+      getRpcMessageHandler:
+        workerExecutionEnvironment.getRpcMessageHandler.bind(
+          workerExecutionEnvironment,
+        ),
+      removeAllPermissionsFor: jest.fn(),
+      getPermissions: jest.fn(),
+      hasPermission: jest.fn(),
+      requestPermissions: jest.fn(),
+      closeAllConnections: jest.fn(),
+      messenger,
+    });
+    const plugin = await pluginController.add({
+      name: 'TestPlugin',
+      sourceCode: `
+        wallet.registerRpcMessageHandler(async (origin, request) => {
+          const {method, params, id} = request;
+          return method + id;
+        });
+      `,
+      manifest: {
+        web3Wallet: {
+          initialPermissions: {},
+        },
+        version: '0.0.0-development',
+      },
+    });
+    await pluginController.startPlugin(plugin.name);
+
+    // defer
+    setTimeout(() => {
+      controllerMessenger.publish(
+        'ServiceMessenger:unhandledError',
+        plugin.name,
+        {
+          message: 'foo',
+          code: 123,
+        },
+      );
+    }, 0);
+
+    await new Promise((resolve) => {
+      controllerMessenger.subscribe(
+        'ServiceMessenger:unhandledError',
+        async () => {
+          const localPlugin = pluginController.get(plugin.name);
+          expect(localPlugin.isRunning).toStrictEqual(false);
+          resolve(undefined);
+        },
+      );
```

Comment:
Non-blocking: For this kind of situation (very common in tests involving events), I think we should standardize on deferred promises, i.e. doing `const promise = new Promise(...)`, then do the thing that causes the promise to resolve or reject, then do `await promise`.